Test fix ke1xf userspace failures

ioannisg opened the pull request:
Add an option to exlude HW Stack protection from user mode tests. This allows us to run the user mode test suite for platforms with limited number of MPU regions, such as the twr_ke18f board.
Fixes #18284 
Fixes #16229
Fixes #16225

ioannisg committed:
soc: arm: ke1xf: do not enable stack prot on tests with user mode.

This commit enabled TEST_USERSPACE_WITHOUT_HW_STACK_PROTECTION
Kconfig option by default for ke1xf SoC Series, which instructs
the build to disable HW stack protection from tests that are to
run with User Mode enabled. This is done because this platform
does not have a sufficient number of MPU regions to support HW
stack protection (Stack Guards) and User Mode simultaneously.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
testsuite: allow disabling stack protection in usermode tests

This commit introduces a new Kconfig option in the testsuite
sub-system, which allows us to disable HW stack protection from
tests that run with user mode enabled.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

henrikbrixandersen commented:
@ioannisg Thank you for working on this. I have verified the fix on my TWR-KE18F board for a few select test cases. I will start a sanitycheck and report back.

henrikbrixandersen commented:
> I would like to understand why ke1xf defines so many hard-coded MPU regions (five) in nxp_mpu_regions.c.

@andrewboie Please see my comment at https://github.com/zephyrproject-rtos/zephyr/issues/16229#issuecomment-495947933

henrikbrixandersen commented:
> @ioannisg Thank you for working on this. I have verified the fix on my TWR-KE18F board for a few select test cases. I will start a sanitycheck and report back.

The following tests still fail due to too few free MPU regions:
- tests/arch/arm/arm_ramfunc/arch.ramfunc
- tests/kernel/mem_protect/mem_protect/kernel.memory_protection
- tests/kernel/mem_protect/obj_validation/kernel.memory_protection.obj_validation
- tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls
- tests/lib/mem_alloc/libraries.libc.newlib

The following tests fail with NXP MPU bus faults:
- samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
- tests/kernel/fatal/kernel.common.stack_protection
- tests/kernel/fatal/kernel.common.stack_protection_arm_fp_sharing

ioannisg committed:
tests: kernel: replace CONFIG_USERSPACE with CONFIG_TEST_USERSPACE

This commit replaces several CONFIG_USERSPACE=y
settings with CONFIG_TEST_USERSPACE=y. This allows
the test sub-system Kconfig structure to control
the settings of USERSPACE and HW_STACK_PROTECTION
in the various tests suites.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?

henrikbrixandersen commented:
> @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?

The following test still fails due to too few free MPU regions:
- tests/lib/mem_alloc/libraries.libc.newlib

The following tests fail with NXP MPU bus faults:
- tests/kernel/fatal/kernel.common.stack_protection
- samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
- tests/kernel/mem_protect/mem_protect/kernel.memory_protection

galak reviewed (changes requested):
Is this because of number of MPU regions?

See comment about where we enable this.

ioannisg commented on soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series:
```diff
@@ -108,6 +108,9 @@ config ADC_MCUX_ADC12
 
 endif # ADC
 
+config TEST_USERSPACE_WITHOUT_HW_STACK_PROTECTION
```
@galak to clarify: you prefer adding `default y if SOC_SERIES_KINETIS_KE1XF` in the original definition of `TEST_USERSPACE_WITHOUT_HW_STACK_PROTECTION `in subsys/testsuite/Kconfig?

ioannisg replied:
that's fine with me, will apply this proposal.

ioannisg replied:
done

ioannisg commented:
> 
> 
> > @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?
> 
> The following test still fails due to too few free MPU regions:
> 
>     * tests/lib/mem_alloc/libraries.libc.newlib
> 
This one is rather simple - we can exclude the platform for this test.

MaureenHelm reviewed (commented):
> I would like to understand why ke1xf defines so many hard-coded MPU regions (five) in nxp_mpu_regions.c.

k64f also uses 5 regions, but it has 12 regions available in hardware whereas ke18f has only 8.

> Does the NXP MPU have no concept of a background System map?

No

> Why do we need an address-space sized region for "DEBUGGER_0"?

To access memory with a debugger.

> Why do we need a region for "RAM_U_0", given that supervisor mode ought to already be able to read/write RAM?
Can we disable "BACKGROUND_0" and "BACKGROUND_1" if CONFIG_HW_STACK_PROTECTION is disabled?

Yes, this may be possible. We should wait to do this until after the 2.0 release because it's really easy to break things when messing with MPU regions. I propose we merge this PR now and open an enhancement issue to optimize the number of boot MPU regions.

> My feeling is that the boot MPU region config for a lot of boards were set up long before userspace was even implemented and need to be revisited.

ioannisg commented:
> 
> 
> > @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?
> 
> The following test still fails due to too few free MPU regions:
> 
>     * tests/lib/mem_alloc/libraries.libc.newlib

This test uses one extra region for Heap, which we do not have, even if we disable stack protection. So I will exclude the platform for this test.
>

ioannisg committed:
tests: exclude twr_ke18f platform from several user mode tests

Several user mode tests cannot run on twr_ke18f because
either the platform does not have a sufficient number of
MPU regions required for the tests, or, the tests also
require HW stack protection (which has been, by default,
excluded in user mode tests for twr_ke18f board). We
excluded the board from all those tests.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
>The following tests fail with NXP MPU bus faults:
>
>    tests/kernel/fatal/kernel.common.stack_protection
>    samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
>    tests/kernel/mem_protect/mem_protect/kernel.memory_protection

These fail because they expect the HW stack protection to function. But we disable it. 
So we need to exclude them from running on this board.

ioannisg commented:
@henrikbrixandersen @MaureenHelm @galak could you take a look one more time, here?

In brief, I applied as little filtering as it was possible, for this platform. Still there are around 30 user mode tests executed for this board, so we keep coverage at decent level.

henrikbrixandersen commented:
> @henrikbrixandersen @MaureenHelm @galak could you take a look one more time, here?
> 
> In brief, I applied as little filtering as it was possible, for this platform. Still there are around 30 user mode tests executed for this board, so we keep coverage at decent level.

I am rerunning sanitycheck against the twr_ke18f board now.
